### PR TITLE
Add support for template HTML based email sending in TOTP

### PR DESCRIPTION
**Pull request by Buddhimah**
## Documentation
To enable this fix we need to add the following configuration under AuthenticatorConfig "totp"
              `   <Parameter name="useEventHandlerBasedEmailSender">true</Parameter>`

## Email Template Documentation

Add following[1] configuration to email-admin-config.xml 
[1] [https://docs.google.com/document/d/1Fhj0u-SW0EbofDvP_Pm3zQCAitiaywF2CYrXA02CFUE/edit?usp=sharing](https://docs.google.com/document/d/1Fhj0u-SW0EbofDvP_Pm3zQCAitiaywF2CYrXA02CFUE/edit?usp=sharing)

 <configuration type="totp" display="TOTP" locale="en_US">
        <targetEpr></targetEpr>
        <subject>WSO2 Carbon - Time-Based One Time Password</subject>
        <body><![CDATA[<table align="center" cellpadding="0" cellspacing="0" border="0" width="100%"bgcolor="#f0f0f0">
        <tr>
                <td style="padding: 30px 30px 20px 30px;">
                        <table cellpadding="0" cellspacing="0" border="0" width="100%" bgcolor="#ffffff" style="max-width: 650px; margin: auto;">
                                <tr>
                                    <td colspan="2" align="center" style="background-color: #333; padding: 40px;">
                                            <a href="http://wso2.com/" target="_blank"></a>
                                    </td>
                                </tr>
                                <tr>
                                    <td colspan="2" align="center" style="padding: 50px 50px 0px 50px;">
                                        <h1 style="padding-right: 0em; margin: 0; line-height: 40px; font-weight:300; font-family: 'Nunito Sans', Arial, Verdana, Helvetica, sans-serif; color: #666; text-align: left; padding-bottom: 1em;">
                                                Time-Based One Time Password
                                        </h1>
                                    </td>
                                </tr>
                                <tr>
                                    <td style="text-align: left; padding: 0px 50px 20px 50px;" valign="top">
                                        <p style="font-size: 18px; margin: 0; line-height: 24px; font-family: 'Nunito Sans', Arial, Verdana, Helvetica, sans-serif; color: #666; text-align: left; padding-bottom: 3%;">
                                                Hi {{user.claim.givenname}},
                                        </p>
                                        <p style="font-size: 18px; margin: 0; line-height: 24px; font-family: 'Nunito Sans', Arial, Verdana, Helvetica, sans-serif; color: #666; text-align: left; padding-bottom: 3%;">
                                                Please use the token <b>{{token}}</b> as the password for your login.
                                        </p>
                                    </td>
                                </tr>
                                <tr>
                                    <td style="text-align: left; padding: 30px 50px 50px 50px" valign="top">
                                        <p style="font-size: 18px; margin: 0; line-height: 24px; font-family: 'Nunito Sans', Arial, Verdana, Helvetica, sans-serif; color: #505050; text-align: left;">
                                             Thanks,<br/>WSO2 Identity Server Team
                                        </p>
                                    </td>
                                </tr>
                                <tr>
                                    <td colspan="2" align="center" style="padding: 20px 40px 40px 40px;" bgcolor="#f0f0f0">
                                        <p style="font-size: 12px; margin: 0; line-height: 24px; font-family: 'Nunito Sans', Arial, Verdana, Helvetica, sans-serif; color: #777;">
                                            &copy; 2018
                                            <a href="http://wso2.com/" target="_blank" style="color: #777; text-decoration: none">WSO2</a>
                                            <br>
                                            787 Castro Street, Mountain View, CA 94041.
                                    </p>
                                    </td>
                                </tr>
                        </table>
                </td>    
        </tr>
</table>]]></body>
        <redirectPath></redirectPath>
    </configuration>

### CODE REVIEW

File: component/authenticator/pom.xml [hunk 1]
```diff
@@ -127,6 +132,7 @@
                             org.wso2.carbon.extension.identity.helper.*;
                             version="${identity.extension.utils.import.version.range}"
                         </Import-Package>
+                        <Import-Package>*;resolution:=optional</Import-Package>
```
Comment (mefarazath) [inline component/authenticator/pom.xml]: Why do we have to do this?

Comment (ruwanta) [inline component/authenticator/pom.xml]: * import is dangerous. You need to import the required package with version range

Comment (denuwanthi) [inline component/authenticator/pom.xml]: why two <Import-Package> elements?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPTokenGenerator.java [hunk 2]
```diff
@@ -119,7 +126,17 @@ public static String generateTOTPTokenLocal(String username, AuthenticationConte
...
-					sendNotification(tenantAwareUsername, firstName, Long.toString(token), email);
+
+					// Check whether the authenticator is configured to use the event handler implementation.
+					if (TOTPUtil.isEventHandlerBasedEmailSenderEnabled()) {
+						AuthenticatedUser authenticatedUser = (AuthenticatedUser) context.getProperty
+								(TOTPAuthenticatorConstants.AUTHENTICATED_USER);
+						triggerEvent(authenticatedUser.getUserName(), authenticatedUser.getTenantDomain(),
+								authenticatedUser.getUserStoreDomain(), TOTPAuthenticatorConstants.EVENT_NAME,
+								Long.toString(token));
+					} else{
```
Comment (mefarazath) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPTokenGenerator.java]: Fix formatting

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPTokenGenerator.java [hunk 3]
```diff
@@ -244,4 +261,34 @@ private static void sendNotification(String tenantAwareUsername, String firstNam
...
+		String eventName = IdentityEventConstants.Event.TRIGGER_NOTIFICATION;
+		HashMap<String, Object> properties = new HashMap<>();
+		properties.put(IdentityEventConstants.EventProperty.USER_NAME, userName);
+		properties.put(IdentityEventConstants.EventProperty.USER_STORE_DOMAIN, userStoreDomainName);
+		properties.put(IdentityEventConstants.EventProperty.TENANT_DOMAIN, tenantDomain);
+		properties.put(TOKEN, otpCode);
+		properties.put(TOTPAuthenticatorConstants.TEMPLATE_TYPE, notificationEvent);
+		Event identityMgtEvent = new Event(eventName, properties);
+		try {
+			System.out.println(TOTPDataHolder.getInstance().getIdentityEventService());
```
Comment (denuwanthi) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPTokenGenerator.java]: why are we using a System.out.println here? can't we use logs?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPTokenGenerator.java [hunk 4]
```diff
@@ -244,4 +261,34 @@ private static void sendNotification(String tenantAwareUsername, String firstNam
...
+			TOTPDataHolder.getInstance().getIdentityEventService().handleEvent(identityMgtEvent);
+		} catch (IdentityEventException e) {
+			String errorMsg = "Error occurred while calling triggerNotification. " + e.getMessage();
```
Comment (denuwanthi) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPTokenGenerator.java]: Let's have more context full eror message. 
1.there is no triggerNotification method.
2.dont use camel case inside the log.
3.lets mention which notification type failed.

**Pull request closed without merging**